pkg/log/logshipper/logshipper.go

```diff
@@ -87,9 +92,7 @@ func (ls *LogShipper) Ping() {
 }
 
 func (ls *LogShipper) Run() error {
-	ctx, cancel := context.WithCancel(context.Background())
-	ls.stopFunc = cancel
-	return ls.sendBuffer.Run(ctx)
+	return ls.sendBuffer.Run(ls.ctx)
```

Comment:
I had to move setting `ls.stopFunc` into `New` to avoid a data race in the test, where `Run` and `Stop` could be reading and setting `ls.stopFunc` at the same time.

Reply:
This feels wrong to me. I'd put a mutex on it if needed.

(It's not high performance, so...)